### PR TITLE
Usage limits on per-code basis

### PR DESCRIPTION
**Pull request by philbirt**
* Promotion eligibility takes into account code usage limits
* Refactor usage limit code in promotion and promotion codes, more specs
* Added specs to promotion handlers for promotion code usage limits
* Admin is able to specify per-code limits as well as top-level promotion limits
* CSV download of codes for admin

Paired w/ @jordan-brough

### CODE REVIEW

**Comment by philbirt** [comment]
cc @gmacdougall @jhawthorn @cbrunsdon This will be the final step, based on the other step we just pushed out (https://github.com/bonobos/spree/pull/258) Mind taking a look?

File: core/app/models/spree/promotion.rb [hunk 1]
```diff
@@ -121,28 +124,36 @@ def eligible_rules(promotable, options = {})
...
-      usage_limit.present? && usage_limit > 0 && adjusted_credits_count(promotable) >= usage_limit
-    end
-
-    def adjusted_credits_count(promotable)
-      credits_count - promotable.adjustments.promotion.where(:source_id => actions.pluck(:id)).count
-    end
-
-    def credits
-      Adjustment.eligible.promotion.where(source_id: actions.map(&:id))
+      # TODO: This logic appears to be wrong.
```
Comment (jordan-brough) [inline core/app/models/spree/promotion.rb]: Note on this: The logic was already like this before this PR.  We just added some comments about it upon noticing it.

**Comment by jordan-brough** [comment]
Awesome!  Left a couple minor comments but :+1: regardless of those.

**Comment by jordan-brough** [comment]
:+1:

File: core/spec/models/spree/promotion_handler/coupon_spec.rb [hunk 2]
```diff
@@ -182,8 +182,23 @@ def expect_adjustment_creation(adjustable:, promotion:, promotion_code:nil)
...
+              order_2.stub :coupon_code => "10off"
+              coupon = Coupon.new(order_2)
+              coupon.apply
+              expect(coupon.successful?).to be false
+              expect(coupon.error).to eq Spree.t(:coupon_code_max_usage)
+            end
+
+
+            it "promotion code hit max usage" do
+              promotion.update!(per_code_usage_limit: 1)
```
Comment (gmacdougall) [inline core/spec/models/spree/promotion_handler/coupon_spec.rb]: Could you tidy up these tests a bit? They're named poorly, don't use contexts well and both the success and failure conditions in the same test.

**Comment by gmacdougall** [comment]
Looking good!

File: core/spec/models/spree/promotion_handler/coupon_spec.rb (hunk 2 above)
Comment (philbirt) [inline core/spec/models/spree/promotion_handler/coupon_spec.rb]: Excellent call, those definitely were hairy. I cleaned up all of the specs around various errors you can encounter when applying a coupon.

**Pull request closed without merging**